third_party/amd/lib/TritonAMDGPUTransforms/RefineOps.cpp

```diff
@@ -636,6 +636,60 @@ LogicalResult rewriteLocalStoreOp(OpBuilder &rewriter,
   return success();
 }
 
+// Reduce ops have different intput and output shapes and produce
+// sliced layouts.
+// This currently only supports 2d inputs.
+LogicalResult rewriteReduceOp(OpBuilder &rewriter, triton::ReduceOp op) {
+  auto ctx = op->getContext();
+  auto loc = op.getLoc();
+  uint32_t axisReduce = op.getAxis();
+  uint32_t axisNonReduce = (axisReduce+1)%2;
+
+  // Calculate refined shape.
+  auto src = op->getOperand(0);
+  auto srcType = rankedTType(src);
+  if (srcType.getRank()!=2)
+    return failure();
+  auto srcShape = srcType.getShape();
+  auto srcEncoding = srcType.getEncoding();
+  auto srcShapePerCtaTile = triton::gpu::getShapePerCTATile(srcEncoding);
+  SmallVector<int64_t> repShape = {srcShape[0] / srcShapePerCtaTile[0], srcShape[1] / srcShapePerCtaTile[1]};
+  int numReps = repShape[axisNonReduce];
+  SmallVector<int64_t> refinedSrcShape = {srcShape[0], srcShape[1]};
+  refinedSrcShape[axisNonReduce] /= numReps;
+  int64_t elementsPerRep = refinedSrcShape[axisNonReduce];
+  auto elemTy = srcType.getElementType();
+  auto refinedTensorType = RankedTensorType::get(refinedSrcShape, elemTy, srcEncoding);
+
+  // Create refined ops.
+  rewriter.setInsertionPointAfter(op);
+  SmallVector<Value> refinedReduces;
+  for (int i = 0; i < numReps; ++i) {
+    SmallVector<int64_t> offset(refinedSrcShape.size(), 0);
+    offset[axisReduce] = 0;
+    offset[axisNonReduce] = i * elementsPerRep;
+    auto sliceOp = rewriter.create<triton::amdgpu::ExtractSliceOp>(
+      loc, Type{refinedTensorType}, Value{src}, offset);
+    auto reduceOp = rewriter.create<triton::ReduceOp>(
+        op.getLoc(), ValueRange{sliceOp}, axisReduce);
```

Comment:
`op.getLoc()` -> `loc` (which is defined at the top of the function)

Reply:
Done.